clang-tools-extra/clang-tidy/modernize/UseScopedLockCheck.cpp

```diff
@@ -0,0 +1,342 @@
+}
+
+// Find the exact source range of the 'lock_guard' token
+SourceRange getLockGuardRange(const TypeSourceInfo *SourceInfo) {
+  const TypeLoc LockGuardTypeLoc = SourceInfo->getTypeLoc();
+
+  return SourceRange(LockGuardTypeLoc.getBeginLoc(),
+                     LockGuardTypeLoc.getEndLoc());
+}
+
+// Find the exact source range of the 'lock_guard' name token
+SourceRange getLockGuardNameRange(const TypeSourceInfo *SourceInfo) {
+  const TemplateSpecializationTypeLoc TemplateLoc =
+      getTemplateLockGuardTypeLoc(SourceInfo);
+  if (!TemplateLoc)
+    return {};
+
+  return SourceRange(TemplateLoc.getTemplateNameLoc(),
+                     TemplateLoc.getLAngleLoc().getLocWithOffset(-1));
+}
+
+AST_MATCHER_P(CompoundStmt, hasMultiple, ast_matchers::internal::Matcher<Stmt>,
+              InnerMatcher) {
+  size_t Count = 0;
+
+  for (const Stmt *Stmt : Node.body())
+    if (InnerMatcher.matches(*Stmt, Finder, Builder))
+      Count++;
+
+  return Count > 1;
+}
+
+AST_MATCHER_P(CompoundStmt, hasSingle, ast_matchers::internal::Matcher<Stmt>,
+              InnerMatcher) {
+  size_t Count = 0;
+  ast_matchers::internal::BoundNodesTreeBuilder Result;
+
+  for (const Stmt *Stmt : Node.body()) {
+    ast_matchers::internal::BoundNodesTreeBuilder TB(*Builder);
+    if (InnerMatcher.matches(*Stmt, Finder, &TB)) {
+      Count++;
+      if (Count == 1)
+        Result.addMatch(TB);
+    }
+  }
+
+  if (Count > 1) {
+    Builder->removeBindings(
+        [](const ast_matchers::internal::BoundNodesMap &) { return true; });
+    return false;
+  }
+
+  *Builder = std::move(Result);
+  return true;
+}
+
+const StringRef UseScopedLockMessage =
+    "use 'std::scoped_lock' instead of 'std::lock_guard'";
+
+} // namespace
+
+UseScopedLockCheck::UseScopedLockCheck(StringRef Name,
+                                       ClangTidyContext *Context)
+    : ClangTidyCheck(Name, Context),
+      WarnOnSingleLocks(Options.get("WarnOnSingleLocks", true)),
+      WarnOnUsingAndTypedef(Options.get("WarnOnUsingAndTypedef", true)) {}
+
+void UseScopedLockCheck::storeOptions(ClangTidyOptions::OptionMap &Opts) {
+  Options.store(Opts, "WarnOnSingleLocks", WarnOnSingleLocks);
+  Options.store(Opts, "WarnOnUsingAndTypedef", WarnOnUsingAndTypedef);
+}
+
+void UseScopedLockCheck::registerMatchers(MatchFinder *Finder) {
+  const auto LockGuardClassDecl =
+      namedDecl(hasName("lock_guard"), isInStdNamespace());
+
+  const auto LockGuardType = qualType(anyOf(
+      hasUnqualifiedDesugaredType(
+          recordType(hasDeclaration(LockGuardClassDecl))),
+      elaboratedType(namesType(hasUnqualifiedDesugaredType(
+          templateSpecializationType(hasDeclaration(LockGuardClassDecl)))))));
+
+  const auto LockVarDecl = varDecl(hasType(LockGuardType));
+
+  if (WarnOnSingleLocks) {
+    Finder->addMatcher(
+        compoundStmt(
+            unless(isExpansionInSystemHeader()),
+            hasSingle(declStmt(has(LockVarDecl.bind("lock-decl-single"))))),
```

Comment:
same here, this single is also usless,
simply do:
has(declStmt(has(LockVarDecl)).bind("first")),
unless(has(declStmt(unless(equalsBoundNode("first"), has(LockVarDecl))))